### PR TITLE
Framework for adding compatibility tests

### PR DESCRIPTION
**Pull request by mcvsubbu**
This PR includes a basic framework for us to build compatibility tests across releases
The actual tests and a lot of code is yet to be written.

The framework is set up so that we can run operations during a cluster upgrade, and
detect incompatibility across releases. The value of the tests will be in the operations
we choose to test during upgrades

Configuration is done via YAML, and some sample yaml files and skinny classes have been
provided.

Issue 4854

## Description
Add a description of your PR here.
A good description should include pointers to an issue or design document, etc.
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release.

If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

File: compatibility-verifier/compCheck.sh [hunk 1]
```diff
@@ -157,20 +181,27 @@ fi
 
 # Setup initial cluster with olderCommit and do rolling upgrade
 startServices "$oldTargetDir"
-sleep 20
+#$COMPAT_TESTER pre-controller-upgrade.yaml
```
Comment (sajjad-moradi) [inline compatibility-verifier/compCheck.sh:184]: I'm not sure why 20 seconds sleep was chosen in the first place, but does replacing it with running compat_tester mean the tests are going to take 20 seconds?
Also maybe it reads better if $COMPAT_TESTER is replaced with a function call, something like runCompatiblityTests, which is more consistent with startService, startService, ... here.

File: compatibility-verifier/compCheck.sh [hunk 2]
```diff
@@ -157,20 +181,27 @@ fi
...
 stopService controller "$newTargetDir"
 startService controller "$oldTargetDir"
+#$COMPAT_TESTER post-server-rollback.yaml
 stopService broker "$newTargetDir"
 startService broker "$oldTargetDir"
+#$COMPAT_TESTER post-broker-rollback.yaml
 stopService server "$newTargetDir"
 startService server "$oldTargetDir"
+#$COMPAT_TESTER post-controller-rollback.yaml
 sleep 20
```
Comment (sajjad-moradi) [inline compatibility-verifier/compCheck.sh:205]: If you decide to have(or remove) previous `sleep`s, please do the same for this one as well?

File: pinot-tools/src/main/java/org/apache/pinot/tools/compat/SegmentOp.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+  public void setTableConfigFileName(String tableConfigFileName) {
+    _tableConfigFileName = tableConfigFileName;
+  }
+
+  @Override
+  boolean runOp() {
+    switch(_op) {
+      case UPLOAD:
+        System.out.println("Generating segment " + _segmentName + " from " + _inputDataFileName + " and uploading to " +
```
Comment (jackjlli) [inline pinot-tools/src/main/java/org/apache/pinot/tools/compat/SegmentOp.java:89]: We may not need to print it out in the console?

**Review by jackjlli — APPROVED** [review]
Minor but LGTM

**Comment by mcvsubbu** [comment]
we should remove all sleeps. They were added by the earlier committer. I will do that. 
We also need to check the status of each compat-tester callout.
Lastly, I think this whole thing needs to move into pinot-integration-tests, so that we can re-use all the code in there in the compat test operations. What do you say?

File: compatibility-verifier/compCheck.sh (hunk 1 above)
Comment (mcvsubbu) [reply compatibility-verifier/compCheck.sh:184]: Removed the sleep. It should not be needed.
Replacing with a function call may be fine, but we still have to check the return from that function. I have added the status check now. Let it be like this for now, until we add more meat to it

File: pinot-tools/src/main/java/org/apache/pinot/tools/compat/SegmentOp.java (hunk 3 above)
Comment (mcvsubbu) [reply pinot-tools/src/main/java/org/apache/pinot/tools/compat/SegmentOp.java:89]: We will not need it. I put it there so we can see what is being run.  Once we add more logic here, we can remove these things

**Review by jackjlli — APPROVED** [review]

**Pull request merged**